Description: Move to the window / move the window to the user depending on Shell or not

=== modified file 'src/empathy-chat-window.c'
Index: empathy-3.12.6/src/empathy-chat-window.c
===================================================================
--- empathy-3.12.6.orig/src/empathy-chat-window.c
+++ empathy-3.12.6/src/empathy-chat-window.c
@@ -67,6 +67,7 @@ struct _EmpathyChatWindowPriv
   GList *chats;
   gboolean page_added;
   gboolean dnd_same_window;
+  gboolean shell_running;
   EmpathyChatroomManager *chatroom_manager;
   EmpathyNotifyManager *notify_mgr;
   EmpathyIndividualManager *individual_mgr;
@@ -172,6 +173,35 @@ static void empathy_chat_window_get_nb_c
 
 G_DEFINE_TYPE (EmpathyChatWindow, empathy_chat_window, GTK_TYPE_WINDOW)
 
+
+#define GNOME_SHELL_BUS_NAME "org.gnome.Shell"
+
+static void
+list_names_cb (TpDBusDaemon *bus_daemon,
+        const gchar * const *names,
+        const GError *error,
+        gpointer user_data,
+        GObject *weak_object)
+{
+  EmpathyChatWindow *self = EMPATHY_CHAT_WINDOW (weak_object);
+  guint i;
+
+  if (error != NULL)
+      goto out;
+
+  for (i = 0; names[i] != NULL; i++)
+    {
+      if (!tp_strdiff (names[i], GNOME_SHELL_BUS_NAME))
+        {
+          self->priv->shell_running = TRUE;
+          break;
+        }
+    }
+
+out:
+  return;
+}
+
 static void
 chat_window_accel_cb (GtkAccelGroup *accelgroup,
     GObject *object,
@@ -2408,6 +2438,8 @@ empathy_chat_window_init (EmpathyChatWin
   GtkWidget *chat_vbox;
   gchar *filename;
   EmpathySmileyManager *smiley_manager;
+  GError *error = NULL;
+  TpDBusDaemon *dbus;
 
   self->priv = G_TYPE_INSTANCE_GET_PRIVATE (self,
     EMPATHY_TYPE_CHAT_WINDOW, EmpathyChatWindowPriv);
@@ -2561,6 +2593,15 @@ empathy_chat_window_init (EmpathyChatWin
       self->priv->chat_manager, "closed-chats-changed",
       G_CALLBACK (chat_window_chat_manager_chats_changed_cb), self);
 
+  /* check if Shell is running */
+  dbus = tp_dbus_daemon_dup (&error);
+  g_assert_no_error (error);
+
+  tp_dbus_daemon_list_names (dbus, -1, list_names_cb,
+   NULL, NULL, G_OBJECT (self));
+
+  g_object_unref (dbus);
+
   chat_window_chat_manager_chats_changed_cb (self->priv->chat_manager,
       empathy_chat_manager_get_num_closed_chats (self->priv->chat_manager), self);
 
@@ -2855,11 +2896,17 @@ empathy_chat_window_present_chat (Empath
 
   empathy_chat_window_switch_to_chat (self, chat);
 
-  /* Don't use tpaw_window_present_with_time () which would move the window
-   * to our current desktop but move to the window's desktop instead. This is
-   * more coherent with Shell's 'app is ready' notication which moves the view
-   * to the app desktop rather than moving the app itself. */
-  empathy_move_to_window_desktop (GTK_WINDOW (self), x_timestamp);
+  /* If gnome-shell is running don't use empathy_window_present_with_time () 
+   * which would move the window to our current desktop but move to the 
+   * window's desktop instead. This is more coherent with Shell's 
+   * 'app is ready' notication which moves the view to the app desktop 
+   * rather than moving the app itself. 
+   * If gnome-shell isn't running, move the window to the current desktop */
+
+  if (!self->priv->shell_running)
+    tpaw_window_present_with_time (GTK_WINDOW (self), x_timestamp);
+  else
+    empathy_move_to_window_desktop (GTK_WINDOW (self), x_timestamp);
 
   gtk_widget_grab_focus (chat->input_text_view);
   return self;
